smart-contracts/contracts/cl-vault/src/vault/deposit.rs

```diff
@@ -46,3 +68,523 @@ pub(crate) fn execute_deposit(
+            sender: recipient,
+        },
+    )?;
+
+    Ok(Response::new().add_submessage(SubMsg::reply_on_success(
+        create_msg,
+        Replies::DepositCreatePosition as u64,
+    )))
+}
+
+pub fn handle_deposit_create_position(
+    deps: DepsMut,
+    env: Env,
+    data: SubMsgResult,
+) -> ContractResult<Response> {
+    let resp: MsgCreatePositionResponse = data.try_into()?;
+    let current_deposit = CURRENT_DEPOSIT.load(deps.storage)?;
+    let bq = BankQuerier::new(&deps.querier);
+    let vault_denom = VAULT_DENOM.load(deps.storage)?;
+
+    // we mint shares according to the liquidity created
+    let liquidity = Decimal::from_str(resp.liquidity_created.as_str())?;
+
+    let total_position = get_position(deps.storage, &deps.querier, &env)?
+        .position
+        .ok_or(ContractError::PositionNotFound)?;
+    let total_liquidity = Decimal::from_str(total_position.liquidity.as_str())?;
+
+    // TODO change error type to something more descriptive
+    let total_shares: Uint128 = bq
+        .supply_of(vault_denom.clone())?
+        .amount
+        .ok_or(ContractError::IncorrectShares)?
+        .amount
+        .parse::<u128>()?
+        .into();
+
+    let ratio = liquidity.checked_div(total_liquidity)?;
+    let user_shares = total_shares.multiply_ratio(ratio.numerator(), ratio.denominator());
+
+    // TODO the locking of minted shares is a band-aid for giving out rewards to users,
+    // once tokenfactory has send hooks, we can remove the lockup and have the users
+    // own the shares in their balance
+    // we mint shares to the contract address here, so we can lock those shares for the user later in the same call
+    let mint = MsgMint {
+        sender: env.contract.address.to_string(),
+        amount: Some(coin(user_shares.u128(), vault_denom).into()),
+        mint_to_address: env.contract.address.to_string(),
+    };
+    // save the shares in the user map
+    LOCKED_SHARES.update(
+        deps.storage,
+        current_deposit.sender.clone(),
+        |old| -> Result<Uint128, ContractError> {
+            if let Some(old_shares) = old {
+                Ok(user_shares + old_shares)
+            } else {
+                Ok(user_shares)
+            }
+        },
+    )?;
+
+    // resp.amount0 and resp.amount1 are the amount of tokens used for the position, we want to refund any unused tokens
+    // thus we calculate which tokens are not used
+    let pool_config = POOL_CONFIG.load(deps.storage)?;
+    let bank_msg = refund_bank_msg(
+        &env,
+        current_deposit,
+        &resp,
+        pool_config.token0,
+        pool_config.token1,
+    )?;
+
+    // merge our position with the main position
+    let fungify = SubMsg::reply_on_success(
+        MsgFungifyChargedPositions {
+            position_ids: vec![total_position.position_id, resp.position_id],
+            sender: env.contract.address.to_string(),
+        },
+        Replies::Fungify.into(),
+    );
```

Comment:
tbh we should just use the dsm routine since its modularized, but lmk what you think.